ipwb/replay.py

```diff
@@ -179,6 +179,126 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def lookup_keys(surt):
+    keyre = re.compile(r"(.+)([,/]).+")
+    key = surt.split("?")[0].strip("/")
+    keys = [key, f"{key}/*"]
+    while "," in key:
+        try:
+            m = keyre.match(key)
+            keys.append(f"{m[1]}{m[2]}*")
+            key = m[1]
+        except Exception as e:
+            key = key.strip("/,")
+
+    return keys
+
+
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, rest = ln.split(maxsplit=1)
+    if key == surtk:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, rest = line.split(maxsplit=1)
+        datetimeK = rest.split()[0].decode()
+        if surtk[-1:] == b'/':
+            surtk = surtk[0:-1]
+
+        # TODO: find a more elegant way for comparison than manually
+        # trimming the last two chars off of the surtk
+
+        if key == surtk:
+            if datetime and datetime != datetimeK:
+                # Rm other close matches, exact found
+                lines.clear()
+                lines.add(line)
+
+                break
+            lines.add(line)
+            # Iterate further to get lines after selection point
+            nextLine = iter.readline()
+            while nextLine:
+                surtk, rest = nextLine.split(maxsplit=1)
+                datetimeK = rest.split()[0].decode()
+                if surtk[-1:] == b'/':
+                    surtk = surtk[0:-1]
+
+                if key == surtk:
+                    lines.add(nextLine)
+
+                nextLine = iter.readline()
+
+            # Continue searching until find first instance
+            right = mid
+        elif key > surtk:
+            left = mid
+        else:
+            right = mid
+
+    # Convert uniq set to list then sort and return
+    ret = sorted(list(lines))
```

Comment:
I am not really happy about holding all the matches in memory and sort them if we can find the first match and iterate from there more efficiently, especially, if a single URI is captured way too many times. However, if it works for now, it's alright.

Reply:
@ibnesayeed In the future, should we:
* find the first match
* count the number of match
* return the `fh` iterator pointing to the first match and count

This would prevent retaining the matched lines. I am open to other approaches.